12-offer-encoding.md

```diff
@@ -0,0 +1,846 @@
+graffiti into the invoice for all to see.
+
+Users can give a tip (or obscure the amount sent) by specifying an
+`invreq_amount` in their invoice request, even though the offer specifies an
+`offer_amount`.  The recipient will only accept this if
+the invoice request amount exceeds the amount it's expecting (i.e. its
+`offer_amount` after any currency conversion, multiplied by `invreq_quantity`, if
+any).
+
+Non-offer-response `invoice_request`s are currently required to
+explicitly state the `invreq_amount` in the chain currency,
+so `offer_amount` and `offer_currency` are redundant (but may be
+informative for the payer to know how the sender claims
+`invreq_amount` was derived).
+
+# Invoices
+
+Invoices are a payment request, and when the payment is made, 
+it can be combined with the invoice to form a cryptographic receipt.
+
+The human-readable prefix for invoices is `lni`.  The recipient can send it in
+response to an `invoice_request` using the `onion_message` `invoice` field
+(in which case, it duplicates the fields including the compulsory
+`invreq_payer_id`), or can be produced spontanously (in which case, it has
+more limited fields, and in particular has no `invreq_payer_id`).
+
+1. `tlv_stream`: `invoice`
+2. types:
+    1. type: 0 (`invreq_metadata`)
+    2. data:
+        * [`...*byte`:`blob`]
+    1. type: 2 (`offer_chains`)
+    2. data:
+        * [`...*chain_hash`:`chains`]
+    1. type: 4 (`offer_metadata`)
+    2. data:
+        * [`...*byte`:`data`]
+    1. type: 6 (`offer_currency`)
+    2. data:
+        * [`...*utf8`:`iso4217`]
+    1. type: 8 (`offer_amount`)
+    2. data:
+        * [`tu64`:`amount`]
+    1. type: 10 (`offer_description`)
+    2. data:
+        * [`...*utf8`:`description`]
+    1. type: 12 (`offer_features`)
+    2. data:
+        * [`...*byte`:`features`]
+    1. type: 14 (`offer_absolute_expiry`)
+    2. data:
+        * [`tu64`:`seconds_from_epoch`]
+    1. type: 16 (`offer_paths`)
+    2. data:
+        * [`...*blinded_path`:`paths`]
+    1. type: 18 (`offer_issuer`)
+    2. data:
+        * [`...*utf8`:`issuer`]
+    1. type: 20 (`offer_quantity_max`)
+    2. data:
+        * [`tu64`:`max`]
+    1. type: 22 (`offer_node_id`)
+    2. data:
+        * [`point`:`node_id`]
+    1. type: 80 (`invreq_chain`)
+    2. data:
+        * [`chain_hash`:`chain`]
+    1. type: 82 (`invreq_amount`)
+    2. data:
+        * [`tu64`:`msat`]
+    1. type: 84 (`invreq_features`)
+    2. data:
+        * [`...*byte`:`features`]
+    1. type: 86 (`invreq_quantity`)
+    2. data:
+        * [`tu64`:`quantity`]
+    1. type: 88 (`invreq_payer_id`)
+    2. data:
+        * [`point`:`key`]
+    1. type: 89 (`invreq_payer_note`)
+    2. data:
+        * [`...*utf8`:`note`]
+    1. type: 160 (`invoice_paths`)
+    2. data:
+        * [`...*blinded_path`:`paths`]
+    1. type: 162 (`invoice_blindedpay`)
+    2. data:
+        * [`...*blinded_payinfo`:`payinfo`]
```

Comment:
Why do we have separate `payment_relay` and `payment_constraints` in `encrypted_data_tlv` for the route blinding proposal, but a single `blinded_payinfo` here?

Reply:
That's because the value here is an aggregated summary for each *whole* blinded path. You compute it by "adding together" (but not using plain addition since there is both a base and proportional fee) all the `payment_relay` for each blinded hop, and taking the most restrictive `payment_constraints`.